File: Graph/Dijkstra.java
import java.util.*;

public class Dijkstra {

    static class Edge {
        int to;
        int weight;

        Edge(int to, int weight) {
            this.to = to;
            this.weight = weight;
        }
    }

    static class Node implements Comparable<Node> {
        int vertex;
        int distance;

        Node(int vertex, int distance) {
            this.vertex = vertex;
            this.distance = distance;
        }

        @Override
        public int compareTo(Node other) {
            return Integer.compare(this.distance, other.distance);
        }
    }

    /**
     * Dijkstra's algorithm - finds shortest path from start to all vertices
     * Time Complexity: O((V + E) log V)
     * Space Complexity: O(V + E)
     *
     * @param graph adjacency list representation of graph
     * @param start starting vertex
     * @return array of shortest distances from start to each vertex
     */
    public static int[] dijkstra(List<List<Edge>> graph, int start) {
        int n = graph.size();
        int[] dist = new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[start] = 0;

        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.offer(new Node(start, 0));

        while (!pq.isEmpty()) {
            Node current = pq.poll();
            int u = current.vertex;
            int d = current.distance;

            // Skip if we already found a better path
            if (d > dist[u]) continue;

            for (Edge edge : graph.get(u)) {
                int v = edge.to;
                int weight = edge.weight;

                if (dist[u] + weight < dist[v]) {
                    dist[v] = dist[u] + weight;
                    pq.offer(new Node(v, dist[v]));
                }
            }
        }

        return dist;
    }
}

File: Graph/Dijkstra.kt
import java.util.*

data class Edge(val to: Int, val weight: Int)

data class Node(val vertex: Int, val distance: Int) : Comparable<Node> {
    override fun compareTo(other: Node): Int = distance.compareTo(other.distance)
}

/**
 * Dijkstra's algorithm - finds shortest path from start to all vertices
 * Time Complexity: O((V + E) log V)
 * Space Complexity: O(V + E)
 *
 * @param graph adjacency list representation of graph
 * @param start starting vertex
 * @return array of shortest distances from start to each vertex
 */
fun dijkstra(graph: List<List<Edge>>, start: Int): IntArray {
    val n = graph.size
    val dist = IntArray(n) { Int.MAX_VALUE }
    dist[start] = 0

    val pq = PriorityQueue<Node>()
    pq.offer(Node(start, 0))

    while (pq.isNotEmpty()) {
        val (u, d) = pq.poll()

        // Skip if we already found a better path
        if (d > dist[u]) continue

        for ((v, weight) in graph[u]) {
            if (dist[u] + weight < dist[v]) {
                dist[v] = dist[u] + weight
                pq.offer(Node(v, dist[v]))
            }
        }
    }

    return dist
}
